### PR TITLE
Combine keyof T inferences

### LINKED ISSUE #21553: generic parameter with default inferred too strictly since 2.7
**TypeScript Version:** 2.8.0-dev.20180201

**Search Terms:** generic parameter default infer keyof 2.7

**Code**

```ts
interface X {
  a: string;
  b: string;
}

function foo<T = X>(x: keyof T, y: keyof T) { }

foo<X>('a', 'b');  // compiles cleanly
foo('a', 'b');  // error TS2345: Argument of type '"b"' is not assignable to parameter of type '"a"'.
```

**Expected behavior:**
Both calls to `foo()` compile with no errors.

**Actual behavior:**
In the second call, when the generic parameter `T` is not set explicitly, an error occurs: `error TS2345: Argument of type '"b"' is not assignable to parameter of type '"a"'`.

It seems that since `2.7`, the generic parameter is tightly inferred from the first parameter to be something like `{a: string}`, so the default (which would have been compatible with both parameters) is not used.

This error does not occur in version `2.6.2`. It does occur in version `2.7.1` and `2.8.0-dev.20180201`.

**Playground Link:**  https://www.typescriptlang.org/play/#src=%0Ainterface%20X%20%7B%0A%20%20a%3A%20string%3B%0A%20%20b%3A%20string%3B%0A%7D%0A%0Afunction%20foo%3CT%20%3D%20X%3E(x%3A%20keyof%20T%2C%20y%3A%20keyof%20T)%20%7B%20%7D%0A%0Afoo%3CX%3E('a'%2C%20'b')%3B%0Afoo('a'%2C%20'b')%3B

**Related Issues:**

### LINKED ISSUE #22376: Regression in Type Inference engine in TypeScript 2.7.1
The following code works great in these TypeScript versions:
  - 2.4.2
  - 2.5.3
  - 2.6.1
  - 2.6.2
  
But it gives a compile error in these TypeScript versions:
  - 2.7.1
  - 2.7.2
  - 2.8.0-dev.20180307

### BugTest.ts

```ts
class Write {
    protected dummy: Write;
}

class Col<s, a> {
    protected dummy: [Col<s, a>, s, a];
}

class Table<Req, Def> {
    protected dummy: [Table<Req, Def>, Req, Def];
}

type MakeTable<T1 extends object, T2 extends object> = {
    [P in keyof T1]: Col<Write, T1[P]>;
} & {
        [P in keyof T2]: Col<Write, T2[P]>;
    };

class ConflictTarget<Cols> {
    public static tableColumns<Cols>(cols: (keyof Cols)[]): ConflictTarget<Cols> {
        return <any>{
            tableCols: cols.slice()
        };
    }

    protected dummy: [ConflictTarget<Cols>, Cols];
}



const bookTable: Table<BookReq, BookDef> = null as any

interface BookReq {
    readonly title: string;
    readonly serial: number;
}

interface BookDef {
    readonly author: string;
    readonly numPages: number | null;
}


function insertOnConflictDoNothing<Req extends object, Def extends object>(_table: Table<Req, Def>, _conflictTarget: ConflictTarget<Req & Def>): boolean {
    throw new Error();
}

function f() {
    insertOnConflictDoNothing(bookTable, ConflictTarget.tableColumns(["serial"]));  // <-- Compile error here
}
```

```
BugTest.ts(49,42): error TS2345: Argument of type 'ConflictTarget<{ serial: any; }>' is not assignable to parameter of type 'ConflictTarget<BookReq & BookDef>'.
  Type '{ serial: any; }' is not assignable to type 'BookReq & BookDef'.
    Type '{ serial: any; }' is not assignable to type 'BookReq'.
      Property 'title' is missing in type '{ serial: any; }'.
```

Manually adding the type argument allows the compile to succeed:

```ts
insertOnConflictDoNothing(bookTable, ConflictTarget.tableColumns<BookReq & BookDef>(["serial"]));
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #21553 kinda. The example will now typecheck (instead of error) as expected, but the inferred type for `T` won't be the default (as it used to be when we made no inference), instead it'll be the combination of all the `keyof T` inferences we made.

This is pretty much the contravariant equivalent of how we now combine mapped type key inferences into a union.

Fixes #22376

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11987,19 +11994,19 @@ namespace ts {
                         // If all inferences were made from contravariant positions, infer a common subtype. Otherwise, if
                         // union types were requested or if all inferences were made from the return type position, infer a
                         // union type. Otherwise, infer a common supertype.
-                        const unwidenedType = context.flags & InferenceFlags.InferUnionTypes || inference.priority & InferencePriority.PriorityImpliesUnion ?
+                        const unwidenedType = context.flags & InferenceFlags.InferUnionTypes || inference.priority & InferencePriority.PriorityImpliesCombination ?
```
Comment (sandersn) [inline src/compiler/checker.ts]: can you extract this into a function similar to `getContravariantInference`? Or naming it something besides `unwidenedType` would make it more obvious why we  check PriorityImpliesCombination and get the union, but then get the intersection if that type is never.

Comment (sandersn) [reply src/compiler/checker.ts]: Looks like it is #22323

Comment (weswigham) [reply src/compiler/checker.ts]: @sandersn how is it now?

**Comment by sandersn** [comment]
Also fixes #21048 since it adds the lower priority for "literal"&rarr;keyof inferences.

**Comment by sandersn** [comment]
I added the test for #22376 to the PR.

**Comment by weswigham** [comment]
@sandersn I've moved the tests

**Pull request merged**